Improve JS sanitize sniffs in PHPCS

migueliriano opened the pull request:
We are receiving many false positives from PHPCS for JavaScript which is properly sanitized using `sanitize` from `DOMPurify`. This is leading to superfluous cutter in our code to disable PHPCS.

lipemat commented on LipematWordpress/Sniffs/Sniff.php:
```diff
@@ -0,0 +1,18 @@
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace LipematWordpress\Sniffs;
```
While the library is `wp-phpcs` these custom sniffs are technically not WordPress specific and could be used on any JS/React project. The `Lipe/Sniffs` namespace follows the pattern of the other "lipemat" libraries.
Also, the directory should be named 'src' not "LipematWordPress".

migueliriano replied:
I realized that phpcs is using the folder name where the `ruleset.xml` is located, for example, if we include the custom sniff with the directory name `src` and run `phpcs -i` to see the list of standard sniffers, you are going to get this:

```
The installed coding standards are MySource, PEAR, PSR1, PSR12, PSR2, Squiz, Zend, WordPress-VIP-Go, WordPressVIPMi
nimum, PHPCompatibility, PHPCompatibilityParagonieRandomCompat, PHPCompatibilityParagonieSodiumCompat, PHPCompatibi
lityWP, VariableAnalysis, WordPress, WordPress-Core, WordPress-Docs, WordPress-Extra and src
```

As you can see in the list, the last name in the list is our standard sniff name `src`, which can be a confusing name, I think we can use the directory `src/Lipe/Sniffs`, that way the standard name that going to be used will be `Lipe`.

Let me know what do you think.

lipemat replied:
👍

lipemat commented on LipematWordpress/Sniffs/Sniff.php:
```diff
@@ -0,0 +1,18 @@
+<?php
+/**
+ * WP phpcs Coding Standard.
+ *
+ * @package lipemat/wp-phpcs
+ * @link https://github.com/lipemat/wp-phpcs
+ * @license https://opensource.org/licenses/MIT MIT
+ */
```
This doesn't really mean anything specific to this project. Is it required?

migueliriano replied:
Actually not, I was just following up on the pattern and comments of others sniffs structure, but we can remove it 👍

lipemat commented on LipematWordpress/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * LipematWordpress_Sniffs_JS_DangerouslySetInnerHTMLSniff.
+ *
+ * @package lipemat\wp-phpcs
+ */
```
This doesn't mean anything to me. Is is required?

lipemat commented on phpcs-sample.xml:
```diff
@@ -94,4 +94,18 @@
+
+    <rule ref="LipematWordpress">
+        <!-- Exclude overridden Sniffs -->
+        <exclude name="WordPressVIPMinimum.JS.DangerouslySetInnerHTML"></exclude>
```
Won't need to exclude this here after the sniffs are complete because we simply won't include `WordPressVIPMinimum.JS`

migueliriano replied:
`Wordpress-VIP-Go` is including all `WordpressVIPMinimum` sniffers, so we have to exclude it, otherwise will be included by that package: https://github.com/Automattic/VIP-Coding-Standards/blob/develop/WordPress-VIP-Go/ruleset.xml#L7

I think we can exclude it in our new custom sniffer using the `ruleset.xml`, that way we don't need to add this `exclude` rule in all the projects that going to be used.

lipemat replied:
We don't want to exclude any sniffs with the `Lipe` ruleset because this library is not project specific and will be conflicting for other non VIP-Go project.
The rule may be excluded per project like so.

```xml
<rule ref="WordPressVIPMinimum.JS">
        <exclude name="WordPressVIPMinimum.JS.DangerouslySetInnerHTML.Found"></exclude>
        <type>error</type>
</rule>
```

Although, once this feature is complete, there will be no reason to include `WordPressVIPMinimum.JS` so we will simply exclude via.

```xml
<rule ref="WordPress-VIP-Go">
        <exclude name="WordPressVIPMinimum.JS" />
</rule>
```

lipemat commented on LipematWordpress/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,116 @@
+
+            $this->escapingFunctions = $this->merge_custom_array(
+                $customEscapeFunctions,
+                $this->escapingFunctions
```
$this->escapingFunctions should default to [ `sanitize` ]

migueliriano committed:
Create custom DangerouslySetInnerHTMLSniff to allow custom scaping

We are receiving many false positives from PHPCS for JavaScript which is
properly sanitized using `sanitize` from `DOMPurify`. This is leading to
superfluous cutter in our code to disable PHPCS.

To improve the sniff for `dangerouslySetInnerHTML` we are creating our
custom sniff.

migueliriano commented on src/Lipe/ruleset.xml:
```diff
@@ -0,0 +1,8 @@
+<ruleset name="Lipe">
+    <description>Custom coding standards to improve sniffers</description>
+    <!-- Exclude overridden Sniffs -->
+    <rule ref="WordPressVIPMinimum">
+        <exclude name="WordPressVIPMinimum.JS.DangerouslySetInnerHTML.Found"></exclude>
+    </rule>
```
With this approach, we are going to continue using the `WordPressVIPMinimum.JS` that hasn't been overridden and we don't need to extend or include it in this project if doesn't need to be modified.

lipemat replied:
This is requiring the entire `WordPressVIPMinimum` ruleset. We definitely don't want the `Lipe` sniff to include any third party sniffs. Any use or excluding of sniffs outside of `Lipe` must be done on a project basis as not all projects use `WordPressVIPMinium`. The example phpcs.xml is a better location for this type of thing.

lipemat commented on src/Lipe/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,112 @@
+     * @var string[]
+     */
+    public $customEscapingFunctions = [
+        'sanitize',
```
`[ 'sanitize']` should replace `$this->escapingFunctions` as the default escaping functions are all PHP based and not necessary to search for. `customEscapingFunctions` is reserved for user added project specific escaping functions and should not have any value by default otherwise the value is overridden when a user supplies a custom value.

lipemat commented on src/Lipe/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,112 @@
+<?php
+namespace Lipe\Sniffs\JS;
```
A good portion of this class is throwing PHPCS warnings.

lipemat replied:
I realized there was no PHPCS config for this project :(. I've added on to the master branch that you may use to fix the PHPCS errors.

lipemat commented on src/Lipe/Sniffs/Sniff.php:
```diff
@@ -0,0 +1,10 @@
+ *
+ * Provides a bootstrap for the sniffs, to reduce code duplication.
+ */
+abstract class Sniff extends \WordPressCS\WordPress\Sniff {
```
Seeing a PHPCS warning here.

lipemat commented on src/Lipe/ruleset.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+<ruleset name="Lipe">
+    <description>Custom coding standards to improve sniffers</description>
```
Change to "Lipe Projects Sniffs & Rule Set"

lipemat commented on phpcs-sample.xml:
```diff
@@ -94,4 +89,6 @@
         </properties>
     </rule>
+
+    <rule ref="Lipe"></rule>
```
Any empty tags should be closed.

lipemat commented on src/Lipe/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * Custom list of functions which escape values for output.
+     **
```
Extra `**` here

lipemat commented on src/Lipe/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,112 @@
+use Lipe\Sniffs\Sniff;
+use PHP_CodeSniffer\Util\Tokens;
+
+/**.
```
Extra line here.

lipemat commented on src/Lipe/Sniffs/JS/DangerouslySetInnerHTMLSniff.php:
```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * Merge custom functions provided via a custom ruleset with the defaults, if we haven't already.
+     **
```
Extra `**` here

lipemat reviewed (changes requested):
Missing the majority of the required sniffs and requested various other changes.

lipemat commented:
@migueliriano This feature requires that all JS sniffs support the `sanitize` escaping function as well as any user supplied custom functions. So far the only sniff which supports `sanitize` is `dangerouslySetInnerHTML`. 

As all the new sniffs will share much of the same logic, it makes sense to:

1. Kill the `Lipe\Sniffs\Sniff` abstract. which contains no logic at this time.
2. Create a new `Lipe\Sniffs\Js_Sniff` abstract which may be shared by all of the `JS` sniffs.

Here is an example file to test against. 

1. All cases should pass PHPCS as written.
2. All cases should fail PHPCS if the `sanitize` call is removed.

```ts
import {sanitize} from 'dompurify';

let arbitrary = 'First &middot; Second';

const Test = ( {} ) => {
	// DangerouslySetInnerHTMLSniff
	return (
		<div>
			<div dangerouslySetInnerHTML={{__html: sanitize( arbitrary )}} />
		</div>
	);

	// HTMLExecutingFunctionsSniff
	$( body ).after( sanitize( arbitrary ) );
	$( body ).append( sanitize( arbitrary ) );
	$( body ).appendTo( sanitize( arbitrary ) );
	$( body ).before( sanitize( arbitrary ) );
	$( body ).html( sanitize( arbitrary ) );
	$( body ).insertAfter( sanitize( arbitrary ) );
	$( body ).insertBefore( sanitize( arbitrary ) );
	$( body ).prepend( sanitize( arbitrary ) );
	$( body ).prependTo( sanitize( arbitrary ) );
	$( body ).replaceAll( sanitize( arbitrary ) );
	$( body ).replaceWith( sanitize( arbitrary ) );
	$( body ).write( sanitize( arbitrary ) );
	$( body ).writeln( sanitize( arbitrary ) );

	// InnerHTMLSniff
	document.getElementById( 'body' ).innerHTML = sanitize( arbitrary );

	// StringConcatSniff
	const str = '<concat>' + sanitize( arbitrary );

	// StrippingTagsSniff
	$( body ).html( sanitize( arbitrary ) ).text();

	// WindowSniff
	window.location.href = sanitize( arbitrary );
	window.location.protocol = sanitize( arbitrary );
	window.location.host = sanitize( arbitrary );
	window.location.hostname = sanitize( arbitrary );
	window.location.pathname = sanitize( arbitrary );
	window.location.search = sanitize( arbitrary );
	window.location.hash = sanitize( arbitrary );
	window.location.username = sanitize( arbitrary );
	window.location.port = sanitize( arbitrary );
	window.location.password = sanitize( arbitrary );
	window.name = sanitize( arbitrary );
	window.status = sanitize( arbitrary );
};
```

migueliriano committed:
Address PR's comment

Added the next sniffers with custom scape functions:

- DangerouslySetInnerHTMLSniff
- HTMLExecutingFunctionsSniff
- InnerHTMLSniff
- StringConcatSniff
- StrippingTagsSniff
- WindowSniff

lipemat commented on phpcs-sample.xml:
```diff
@@ -12,9 +12,9 @@
     <file>./wp-content/themes/core</file>
 
     <!-- sniff both PHP and JS -->
-    <arg name="extensions" value="php/PHP,js/JS,jsx/JS,ts/JS,tsx/JS" />
```
A general note about recent code reviews:
I've noticed there are a lot of files which have a whole bunch of whitespace changes. Most likely an auto format on save in your IDE. While it's nice to clean up a files formatting, especially during refactor, changing like every line in a file make a code review quite verbose and destroys the Git history for the file.

For instance, in this file you changed like 3 lines of consequence but yet the Git history would show you changed 61 lines which makes it hard to know what was changed for this commit. 

Please update this commit to only include the changed lines in this file. 

Generally moving forward, unless you are doing a refactor, please only change the lines of consequence and/or the few lines surrounding the lines of consequence. If you IDE is set to "Reformat code" on "Actions on Save", you'll want to use "Changed lines" instead of "Whole file"

lipemat commented on phpcs-sample.xml:
```diff
@@ -25,73 +25,72 @@
         </properties>
     </rule>
+
+    <rule ref="Lipe">
```
The while `Lipe` space might not be errors. Let's be more specific.
```suggestion
    <rule ref="Lipe.JS">
```

lipemat committed:
Introduce phpcs.xml.dist file

The custom ruleset for the WP PHPCS itself.

lipemat committed:
Git ignore the PHPCS cache file

.phpcs.cache

lipemat commented on src/Lipe/Sniffs/JS/StringConcatSniff.php:
```diff
@@ -0,0 +1,42 @@
+     */
+    public function process_token($stackPtr)
+    {
+        $nextToken = $this->phpcsFile->findNext(Tokens::$emptyTokens, $stackPtr + 1, null, true, null, true);
```
Receiving a PHPCS error when testing this case.
```
// StringConcatSniff
	const str = '<concat>' + sanitize( arbitrary );
```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 32 | ERROR | HTML string concatenation detected, this is a security risk, use DOM node construction or a templating language instead: sanitize+.
    |       | (Lipe.JS.StringConcat.Found)
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

```

migueliriano replied:
Actually, this is working correctly, in this case, we have to use the `sanitize` for the `HTML` element instead of the string, `StringConcat` it's looking for HTML tags, not for a string in the validation.

See: https://github.com/Automattic/VIP-Coding-Standards/blob/develop/WordPressVIPMinimum/Sniffs/JS/StringConcatSniff.php#L52

I know that the error message is pointing out the `string` instead of the `HTML` element and can be confusing, but that's the way that they handle the message 😕.

This is an example code that you can use for testing:

```
// StringConcatSniff
	const str = 'test' + sanitize( '<concat>' ) + 'test' + sanitize( '</concat>' );
```
I don't know if it is a bug, but when the second condition is true, they are sending the message of the `$nextToken` instead of `$prevToken`: https://github.com/Automattic/VIP-Coding-Standards/blob/develop/WordPressVIPMinimum/Sniffs/JS/StringConcatSniff.php#L60

For that reason, the error message is showing a string when the HTML element is before the string concat.

lipemat commented on src/Lipe/Sniffs/JS/InnerHTMLSniff.php:
```diff
@@ -0,0 +1,35 @@
+        $functionToken = $this->phpcsFile->findNext(Tokens::$functionNameTokens, $stackPtr + 1);
+
+        if ($this->isScapeFunction($functionToken)) {
+            // It's a scape function.
```
All of these `// It's a scape function` comments may be removed. The method is named the same as the comment so they are redundant.

lipemat commented on src/Lipe/Traits/EscapeOutputFunctions.php:
```diff
@@ -0,0 +1,69 @@
+         *
+         * @var array
+         */
+        $this->addedCustomFunctions = [];
```
This should be added to the trait as properties instead inside a __construct

lipemat commented on src/Lipe/Traits/EscapeOutputFunctions.php:
```diff
@@ -0,0 +1,69 @@
+     */
+    protected function mergeFunctionLists()
+    {
+        $defaultScapeFunctions = ['sanitize' => true];
```
Lets set trait property `escapingFunctions` to `['sanitize' => true]` which will override the `\WordPressCS\WordPress\Sniff::$escapingFunctions` property and prevent accidentally parsing a bunch of PHP escaping functions.

lipemat commented on src/Lipe/Traits/EscapeOutputFunctions.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        $defaultScapeFunctions = ['sanitize' => true];
+
+        if ($this->customEscapingFunctions !== $this->addedCustomFunctions) {
```
Lets follow the same `$this->addedCustomFunctions['escape']` pattern we have in `\WordPressCS\WordPress\Sniffs\Security\EscapeOutputSniff::mergeFunctionLists`

```php
if ( $this->customEscapingFunctions !== $this->addedCustomFunctions['escape'] ) {
			$customEscapeFunctions = $this->merge_custom_array( $this->customEscapingFunctions, array(), false );

			$this->escapingFunctions = $this->merge_custom_array(
				$customEscapeFunctions,
				$this->escapingFunctions
			);

			$this->addedCustomFunctions['escape'] = $this->customEscapingFunctions;
		}
```

lipemat commented on src/Lipe/ruleset.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0"?>
+<ruleset name="Lipe">
+    <description>Lipe Projects Sniffs &amp; Rule Set</description>
```
```suggestion
    <description>Lipe Project - Rules</description>

```

lipemat committed:
Exclude WordPress.NamingConventions.ValidFunctionName.MethodNameInvalid in PHPCS

lipemat commented on src/Lipe/Traits/EscapeOutputFunctions.php:
```diff
@@ -0,0 +1,69 @@
+        $this->addedCustomFunctions = [];
+    }
+
+    protected function isScapeFunction($functionToken)
```
```suggestion
    protected function isEscapeFunction($functionToken)

```

lipemat reviewed (changes requested):
Getting much closer. :)

migueliriano committed:
Merge remote-tracking branch 'origin/master' into develop

migueliriano committed:
Addressed PR's comment

- Fix phpcs errors
- Clean up comments
- Use correct flow for `EscapeOutputFunctions::mergeFunctionLists`
